July-Morning/assignment2/Assignment_2.cpp

```diff
@@ -0,0 +1,166 @@
+#include <iostream>
+#include <string>
+#include <list>
+
+template <typename T> struct node
+{
+    T key;
+    node * left;
+    node * right;
+    node (T init_key)
+    {
+        key = init_key;
+        left = NULL;
+        right = NULL;
+    }
+};
+
+// Here we assume that all keys in the tree are unique
+template <typename T> class binary_tree 
+{
+    public:
+    
+        node<T> * root;
+    
+        binary_tree() : root(NULL) {}
+        
+        ~binary_tree()
+        {
+            if (root != NULL)
+                destroy(root);
+        }
+        
+        void destroy (node<T> * s_root) 
+        {
+            if (s_root->left != NULL)
+                destroy (s_root->left);
+            if (s_root->right != NULL)
+                destroy (s_root->right);
+            delete s_root;   
+        }
+        
+        bool is_child (T& key, node<T> * parent) 
+        {
+            if (parent == NULL)
+                return false;
+            if (parent->left != NULL)    
+                if (parent->left->key == key)
+                    return true;
+            if (parent->right != NULL)    
+                if (parent->right->key == key)
+                    return true;        
+            return (is_child(key, parent->left) || is_child(key, parent->right));
+        }
+        
+        // Find ancestors of the node with a given key
+        // Return value:
+        // true - if ancestors were found successfully
+        // false - if tree is empty or node with given key does not exist
+        bool find_ancestors (node<T> * s_root, T& key, std::list<node<T> *>& ancestors) 
+        {
+            if (s_root == NULL)
+                return false;
+            if (s_root->key == key)
+                return true;
+            bool is_child = (find_ancestors(s_root->left, key, ancestors) || find_ancestors(s_root->right, key, ancestors));    
+            if (is_child)
+                ancestors.push_back(s_root);  
+            return is_child;    
+        }
+        
+        // Find lowest common ancestor of the nodes with given keys
+        // Return value:
+        // true - if ancestor was found successfully
+        // false - if tree is empty or node with at least one of the given keys does not exist
+        // ...or one of the given keys is at the root of the tree
+        bool find_lowest_common_ancestor (node<T> * s_root, T& key_1, T& key_2, T& lca_key) 
+        {
+            if (s_root == NULL)
+                return false;
+            bool left = find_lowest_common_ancestor (s_root->left, key_1, key_2, lca_key);
+            if (left)
+                return true;
+            bool right = find_lowest_common_ancestor (s_root->right, key_1, key_2, lca_key);
+            if (right)
+                return true;
+            if (!((is_child (key_1, s_root)) && (is_child (key_2, s_root))))
```

Comment:
Again, you can spare some of the parenthesis to give a cleaner look:
```C++
if (!(is_child(key_1, s_root) && is_child(key_2, s_root)))
```
You can use De Morgan's laws to even further simplify this if:
```C++
if (!is_child(key_1, s_root) || !is_child(key_2, s_root))
// Easy to read: if one of the keys is not in the tree, return false.
```
Another option is to flip the true and false cases:
```C++
if (is_child(key_1, s_root) && !is_child(key_2, s_root)) {
    // Both keys are in the tree, common ancestor found!
    lca_key = s_root->key;
    return true;
}
return false;
```
The latter is the easiest to read in my opinion because it has the least 'negatives'.
("found both keys ==> common ancestor" rather than "a key was not found ==> not an ancestor")